PR: https://github.com/odoo/odoo/pull/79699

From: 3ae12a6f25af6484a3412b1e63ef17283ba904f0
From: Philippe Wauthy
Date: 2021-11-30 13:41:31

Structural Changes: 5
Total Changes: 29

[IMP] hr_holidays: add default filters to smart buttons

Add default filters to the Allocations and Time Off smartbuttons on the time off type form.
This is done to avoid having numerous records with time and only show the relevant to approved and approved allocations as well as the time off for the current year.

task-2688122

closes odoo/odoo#79699

Signed-off-by: Kevin Baptiste <kba@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr_holidays/models/hr_leave_type.py
+++ b/addons/hr_holidays/models/hr_leave_type.py
@@ -347,13 +347,15 @@ class HolidaysType(models.Model):
             holiday_status.closest_allocation_to_expire = result.get('closest_allocation_to_expire', 0)
 
     def _compute_allocation_count(self):
-        date_from = fields.Datetime.to_string(datetime.datetime.now().replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0))
+        min_datetime = fields.Datetime.to_string(datetime.datetime.now().replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0))
+        max_datetime = fields.Datetime.to_string(datetime.datetime.now().replace(month=12, day=31, hour=23, minute=59, second=59))
         domain = [
             ('holiday_status_id', 'in', self.ids),
-            '|',
-            ('date_from', '>=', date_from),
-            ('date_from', '=', False),
+            ('date_from', '>=', min_datetime),
+            ('date_from', '<=', max_datetime),
+            ('state', 'in', ('confirm', 'validate')),
         ]
+
         grouped_res = self.env['hr.leave.allocation'].read_group(
             domain,
             ['holiday_status_id'],
@@ -364,9 +366,16 @@ class HolidaysType(models.Model):
             allocation.allocation_count = grouped_dict.get(allocation.id, 0)
 
     def _compute_group_days_leave(self):
+        min_datetime = fields.Datetime.to_string(datetime.datetime.now().replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0))
+        max_datetime = fields.Datetime.to_string(datetime.datetime.now().replace(month=12, day=31, hour=23, minute=59, second=59))
+        domain = [
+            ('holiday_status_id', 'in', self.ids),
+            ('date_from', '>=', min_datetime),
+            ('date_from', '<=', max_datetime),
+            ('state', 'in', ('validate', 'validate1', 'confirm')),
+        ]
         grouped_res = self.env['hr.leave'].read_group(
-            [('holiday_status_id', 'in', self.ids),
-             ('date_from', '>=', fields.Datetime.to_string(datetime.datetime.now().replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0)))],
+            domain,
             ['holiday_status_id'],
             ['holiday_status_id'],
         )
@@ -429,13 +438,12 @@ class HolidaysType(models.Model):
                 datetime.datetime.now().replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0))
         action['domain'] = [
             ('holiday_status_id', 'in', self.ids),
-            '|',
-            ('date_from', '>=', date_from),
-            ('date_from', '=', False),
         ]
         action['context'] = {
             'default_holiday_type': 'department',
             'default_holiday_status_id': self.ids[0],
+            'search_default_approved_state': 1,
+            'search_default_year': 1,
         }
         return action
 
@@ -444,10 +452,11 @@ class HolidaysType(models.Model):
         action = self.env["ir.actions.actions"]._for_xml_id("hr_holidays.hr_leave_action_action_approve_department")
         action['domain'] = [
             ('holiday_status_id', '=', self.ids[0]),
-            ('date_from', '>=', fields.Datetime.to_string(datetime.datetime.now().replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0)))
         ]
         action['context'] = {
             'default_holiday_status_id': self.ids[0],
+            'search_default_need_approval_approved': 1,
+            'search_default_this_year': 1,
         }
         return action
 
